### PR TITLE
Merging #5342 (IN with function result).

### PR DESCRIPTION
**Pull request by CurtizJ**
hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
Improvement

Short description (up to few sentences):
Allow constant functions and constant arrays to be used on the right side of IN operator.

Detailed description (optional):
Any constant function returning single value, tuple or array of compatible types can now be used on the right side of the IN operator.

Merging #5342.

### CODE REVIEW

**Comment by CurtizJ** [comment]
Currently have significant performance degradation.
Benchamark on query:
`SELECT count() FROM numbers(10000000) WHERE number IN (... set of 1000 numbers ...)`

19.17
```
localhost:9000, queries 281, QPS: 15.339, RPS: 153806048.025, MiB/s: 1173.447, result RPS: 15.339, result MiB/s: 0.000.

0.000%          0.062 sec.
10.000%         0.063 sec.
20.000%         0.064 sec.
30.000%         0.064 sec.
40.000%         0.064 sec.
50.000%         0.065 sec.
60.000%         0.065 sec.
70.000%         0.066 sec.
80.000%         0.066 sec.
90.000%         0.067 sec.
95.000%         0.068 sec.
99.000%         0.076 sec.
99.900%         0.078 sec.
99.990%         0.078 sec.
```

this PR:
```
localhost:9000, queries 152, QPS: 9.206, RPS: 92313548.463, MiB/s: 704.296, result RPS: 9.206, result MiB/s: 0.000.

0.000%          0.103 sec.
10.000%         0.105 sec.
20.000%         0.106 sec.
30.000%         0.106 sec.
40.000%         0.107 sec.
50.000%         0.108 sec.
60.000%         0.109 sec.
70.000%         0.110 sec.
80.000%         0.111 sec.
90.000%         0.113 sec.
95.000%         0.115 sec.
99.000%         0.119 sec.
99.900%         0.121 sec.
99.990%         0.121 sec.
```

**Comment by CurtizJ** [comment]
Build from the original PR shows the same performance as lastet releases. But when I rebased it on master, performance has been degraded even without commit 0a8080e. I suppose that function `convertFieldToType` or `evaluateConstantExpression` has been slowed down since `May 19, 2019`

**Comment by CurtizJ** [comment]
Benchmark results below are irrelevant. My mistake was that I have comparing release gcc build against my locally builded release clang build. With builds on the same compilers performace is alomost identical. I will add a performance test later.

File: dbms/tests/queries/0_stateless/01071_in_array.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+select [1, 2] in [1, 2];
+select (1, 2) in (1, 2);
+select (1, 2) in [(1, 3), (1, 2)];
+select [1] in [[1], [2, 3]];
+select NULL in NULL;
+select ([1], [2]) in ([NULL], [NULL]);
+select ([1], [2]) in (([NULL], [NULL]), ([1], [2]));
+select ([1], [2]) in [([NULL], [NULL]), ([1], [2])];
```
Comment (Enmk) [inline dbms/tests/queries/0_stateless/01071_in_array.sql]: Would that make sense to  also check single elements (and single element-tuples/arrays)?:
```
select 1 in [1, 2];
select 3 in [1, 2];
select 1 in 1;
select 3 in 1;
select (1) in [1, 2];
select (3) in [1, 2];
select [1] in [1, 2];
select [3] in [1, 2];
```

Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/01071_in_array.sql]:
```
select [1] in [1, 2];
select [3] in [1, 2];
```

Not this case.

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/tests/queries/0_stateless/01071_in_array.sql (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/01071_in_array.sql]: Negative tests can be added...

Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/01071_in_array.sql]: I've added this test, thank you!

**Pull request merged**